### PR TITLE
Implement new diagnostics system

### LINKED ISSUE #351: Support cancellation for requests
Using cancellation tokens we can cancel requests if they are no longer needed. Since most code is synchronous, we will need to use file based cancellation pipes (this is was Typescript does) added here: https://github.com/microsoft/vscode-languageserver-node/pull/587.

### PR DESCRIPTION
**Pull request by jmbockhorst**
Implementation of the new diagnostics system with support for cancellation. Early testing is showing a dramatic increase in responsiveness of completions, diagnostics, etc on large workspaces. I enabled type inference diagnostics for now. While these are not ready for general use, they do work on some workspaces (Rare parsing bugs or type inference bugs can cause issues). We can remove these before the next release if they aren't stable enough or enabled them under a "experimental" setting. I will be testing this at work over the next few days to hopefully iron out any issues. Closes #351.

Here is a summary of the changes:
* Upgraded to `@next` version of language server protocol to support file-based cancellation tokens
* Add interruptable diagnostic provider
* Enable type inference compiler diagnostics
* Removed elm-analyse
* Expose the server protocol through the package

Client PR: https://github.com/elm-tooling/elm-language-client-vscode/pull/135

### CODE REVIEW

**Comment by Razzeee** [comment]
This is conflicting btw

File: src/index.ts [hunk 1]
```diff
@@ -86,6 +90,6 @@ connection.listen();
 // Don't die on unhandled Promise rejections
 process.on("unhandledRejection", (reason, p) => {
   connection.console.error(
-    `Unhandled Rejection at: Promise ${p} reason:, ${reason}`,
+    `Unhandleds Rejection at: Promise ${p} reason:, ${reason}`,
```
Comment (Razzeee) [inline src/index.ts:93]: Typo?

File: src/providers/diagnostics/elmMakeDiagnostics.ts [hunk 2]
```diff
@@ -112,11 +112,11 @@ export class ElmMakeDiagnostics {
     { moduleName: string; valueName?: string; diagnostic: Diagnostic }[]
   >();
   private settings: Settings;
-  private connection: IConnection;
+  private connection: Connection;
 
   constructor() {
     this.settings = container.resolve("Settings");
-    this.connection = container.resolve<IConnection>("Connection");
+    this.connection = container.resolve<Connection>("Connection");
```
Comment (Razzeee) [inline src/providers/diagnostics/elmMakeDiagnostics.ts:119]: What's the reason for this change?

File: src/util/settings.ts [hunk 3]
```diff
@@ -28,15 +29,12 @@ export class Settings {
...
+  private connection: Connection;
 
   private initDone = false;
 
-  constructor(
-    config: IClientSettings,
-    private clientCapabilities: ClientCapabilities,
-  ) {
-    this.connection = container.resolve<IConnection>("Connection");
+  constructor(config: any, private clientCapabilities: ClientCapabilities) {
```
Comment (Razzeee) [inline src/util/settings.ts:36]: Do we really need to go back to `any` here?

File: tsconfig.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "compilerOptions": {
-    "baseUrl": ".",
+    // "baseUrl": ".",
```
Comment (Razzeee) [inline tsconfig.json:3]: Is this meant to still be commented out? Probably an oversight?

File: src/providers/documentFormatingProvider.ts [hunk 5]
```diff
@@ -69,6 +75,7 @@ export class DocumentFormattingProvider {
       }
 
       const settings = await this.settings.getClientSettings();
+      console.log("formatting text");
```
Comment (Razzeee) [inline src/providers/documentFormatingProvider.ts:78]: This is probably something you want to remove?

File: src/providers/diagnostics/elmMakeDiagnostics.ts (hunk 2 above)
Comment (jmbockhorst) [reply src/providers/diagnostics/elmMakeDiagnostics.ts:119]: The language server changed. Maybe the interface moved somewhere else.

File: src/providers/astProvider.ts [hunk 6]
```diff
@@ -55,6 +61,7 @@ export class ASTProvider {
     params: DidChangeTextDocumentParams | DidOpenTextDocumentParams,
     elmWorkspace: IElmWorkspace,
   ): void => {
+    console.log(`handleChangeTextDocument: ${Date.now()}`);
```
Comment (Razzeee) [inline src/providers/astProvider.ts:64]: Remove?

File: src/providers/astProvider.ts [hunk 7]
```diff
@@ -115,12 +125,17 @@ export class ASTProvider {
         true,
       );
 
+      console.log(`Parsed tree ${(tree as any)[0]}`);
```
Comment (Razzeee) [inline src/providers/astProvider.ts:128]: Remove?

File: tsconfig.json (hunk 4 above)
Comment (jmbockhorst) [reply tsconfig.json:3]: It was actually intentional. Since we don't rewrite import paths, it creates auto imports that don't work. I think we need to get a path rewrite module if we want that to work.

File: src/providers/astProvider.ts [hunk 8]
```diff
@@ -3,36 +3,42 @@ import { container } from "tsyringe";
...
 
   constructor() {
     this.parser = container.resolve("Parser");
-    this.connection = container.resolve<IConnection>("Connection");
-    this.documentEvents = container.resolve<TextDocumentEvents>(
-      TextDocumentEvents,
-    );
+    this.connection = container.resolve("Connection");
+    this.documentEvents = container.resolve(TextDocumentEvents);
+    this.clientSettings = container.resolve("ClientSettings");
```
Comment (Razzeee) [inline src/providers/astProvider.ts:41]: These don't seem to be used?

File: src/providers/diagnostics/typeInferenceDiagnostics.ts [hunk 9]
```diff
@@ -8,64 +9,146 @@ import {
...
+              severity: DiagnosticSeverity.Error,
               source: this.TYPE_INFERENCE,
-            };
+            });
           }
-        })
-        .filter(Utils.notUndefined.bind(this));
+        });
+    } catch (error) {
+      //
```
Comment (Razzeee) [inline src/providers/diagnostics/typeInferenceDiagnostics.ts:127]: Should we log something?

File: src/util/delayer.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
+export interface ITask<T> {
+  (): T;
+}
+
+export class Delayer<T> {
+  public defaultDelay: number;
+  private timeout: any; // Timer
```
Comment (Razzeee) [inline src/util/delayer.ts:7]: Can we use a better type?

File: src/util/multistepOperation.ts [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+      if (this.cancellationToken?.isCancellationRequested) {
+        stop = true;
+      } else {
+        action(this);
+      }
+    } catch (e) {
+      stop = true;
+      // ignore cancellation request
+      if (!(e instanceof OperationCanceledException)) {
+        console.log(e, `delayed processing of request`);
```
Comment (Razzeee) [inline src/util/multistepOperation.ts:67]: I guess logging to the connection would be inconvenient here?

File: src/cancellation.ts [hunk 12]
```diff
@@ -0,0 +1,214 @@
+/* eslint-disable @typescript-eslint/no-use-before-define */
```
Comment (Razzeee) [inline src/cancellation.ts:1]: Are these now files an original implementation or are they based on something else? If that's the case, it might make sense to link to them for reference?

File: src/providers/diagnostics/diagnosticsProvider.ts [hunk 13]
```diff
@@ -128,81 +145,95 @@ export class DiagnosticsProvider {
...
+        if (writeable) {
+          this.pendingDiagnostics.set(uri, Date.now());
         }
       });
     });
+
+    try {
+      this.triggerDiagnostics();
+    } catch (e) {
+      console.log(e);
```
Comment (Razzeee) [inline src/providers/diagnostics/diagnosticsProvider.ts:185]: Should probably log to the connection?

File: tsconfig.json (hunk 4 above)
Comment (Razzeee) [reply tsconfig.json:3]: What I'm trying to say is, either remove completely or document why it's commented out

File: src/providers/diagnostics/diagnosticsProvider.ts (hunk 13 above)
Comment (jmbockhorst) [reply src/providers/diagnostics/diagnosticsProvider.ts:185]: I think I just used it for testing. Shouldn't need it anymore.

File: src/providers/diagnostics/elmMakeDiagnostics.ts (hunk 2 above)
Comment (jmbockhorst) [reply src/providers/diagnostics/elmMakeDiagnostics.ts:119]: Yeah, IConnection went away. It was just an empty interface that extended Connection.

File: src/providers/diagnostics/typeInferenceDiagnostics.ts (hunk 9 above)
Comment (jmbockhorst) [reply src/providers/diagnostics/typeInferenceDiagnostics.ts:127]: We shouldn't need it anymore.

File: src/cancellation.ts (hunk 12 above)
Comment (jmbockhorst) [reply src/cancellation.ts:1]: Added links in multiple places.

File: tsconfig.json (hunk 4 above)
Comment (jmbockhorst) [reply tsconfig.json:6]: Added some comments to clarify.

**Comment by Razzeee** [comment]
LGTM, let's fix the tests and get this merged.
